lib/features/support/resilient_driver.rb

```diff
@@ -0,0 +1,106 @@
+require 'appium_lib'
+require_relative './fast_selenium'
+require_relative './logger'
+
+# Handles Appium driver restarts and retries in the case of failure
+class ResilientAppiumDriver < AppAutomateDriver
```

Comment:
I'm not sure how the `AppAutomateDriver` is really used, but this feels like this is a place where `method_missing` would be useful. It would let us forward all calls onto the `AppAutomateDriver` without having to write replacements for each method individually. Looking at the docs, it seems like that approach would work fine and it'd mean we wouldn't have to add new methods to both `AppAutomateDriver` & `ResilientAppiumDriver`

e.g.

```ruby
class ResilientAppiumDriver
  def initialize
    @driver = AppAutomateDriver.new
  end

  def method_missing(method, *args, &block)
    raise "Method '#{method}' does not exist" unless @driver.respond_to?(method)

    begin
      @driver.send(method, *args, &block)
    rescue
      recover
      method_missing(method, *args, &block)
    end
  end

  def recover
    # ...
  end
end
```

I think it'd also be good to cap the number of retries to some arbitrary limit. I know timeouts will eventually kill it, but I'm not sure it makes sense to retry infinitely until then as it could delay knowing about actual problems — e.g. Travis' timeout is >1 hour, so it's not a trivial amount of time

```ruby
class ResilientAppiumDriver
  def initialize
    @driver = AppAutomateDriver.new
  end

  def method_missing(method, *args, &block)
    raise "Method '#{method}' does not exist" unless @driver.respond_to?(method)

    retries = 0

    until retries >= 5 # or some other arbitrary number
      begin
        @driver.send(method, *args, &block)

        return
      rescue => error
        retries += 1

        recover
      end
    end

    # re-raise the last error, so we don't hide the actual failure
    # it might be better to re-raise the _first_ error instead?
    # I'm not sure how likely it is that they will be different
    raise error
  end

  def recover
    # ...
  end
end
```

Reply:
Nice, TIL.  And thanks for the code samples to give me a running start.  I'll give it a go and rerun my soak test for the Cocoa repo.
For info, `AppAutomateDriver` is instantiated in the notifier repos (`env.rb`) and opting into this just means changing the class that's used.

Reply:
Something I forgot about is the docs won't list any methods for `ResilientAppiumDriver` because of `method_missing`

I can't see a way to automatically include all of the methods from `AppAutomateDriver`, but you can [manually document each method if you want](https://rubydoc.info/gems/yard/file/docs/Tags.md#method), or just link to `AppAutomateDriver` with `@see AppAutomateDriver`

Reply:
I vote `@see AppAutomateDriver`